polygeist/test/polygeist-opt/parallellicm.mlir

```diff
@@ -1,108 +1,190 @@
-      func.call @use(%v) : (f32) -> ()
-    }
-    return
+  %a = memref.alloca() : memref<f32>
+  memref.store %arg1, %a[] : memref<f32>
+  affine.for %arg2 = 0 to 10 {    
+    %tmp = memref.load %a[] : memref<f32>
+    %arr = affine.load %arg0[0] : memref<?xf32>
+    %add = arith.addf %arr, %tmp : f32
+    affine.store %add, %arg0[0] : memref<?xf32>
   }
+  return
+}
 }
 
 // -----
 
 module {
-  func.func private @use(f32) 
-  func.func @affhoist(%arg0: memref<?xf32>, %arg1: index, %arg2: index, %arg3 : index, %arg4 : index, %arg5: index, %arg6: index) {
-    // CHECK:   func.func @affhoist(%arg0: memref<?xf32>, %arg1: index, %arg2: index, %arg3: index, %arg4: index, %arg5: index, %arg6: index) {
-    // CHECK-NEXT:     %cst = arith.constant 0.000000e+00 : f32
-    // CHECK-NEXT:     %c1 = arith.constant 1 : index
-    // CHECK-NEXT:     %alloca = memref.alloca() : memref<f32>
-    // CHECK-NEXT:     memref.store %cst, %alloca[] : memref<f32>
-    // CHECK-NEXT:     affine.if #set(%arg1, %arg2, %arg5, %arg3, %arg4, %arg6) {
-    // CHECK-NEXT:       %0 = memref.load %alloca[] : memref<f32>
-    // CHECK-NEXT:       affine.parallel (%arg7, %arg8) = (max(%arg1, %arg2), %arg5) to (min(%arg3, %arg4), %arg6) {
-    // CHECK-NEXT:         func.call @use(%0) : (f32) -> ()
-    // CHECK-NEXT:       }
-    // CHECK-NEXT:     }
-    %cst = arith.constant 0.000000e+00 : f32
-    %c1 = arith.constant 1 : index 
-    %a = memref.alloca() : memref<f32>
-    memref.store %cst, %a[] : memref<f32>
-    affine.parallel (%arg7, %arg8) = (max(%arg1, %arg2), %arg5) to (min(%arg3, %arg4), %arg6) {
-      %v = memref.load %a[] : memref<f32>
-      func.call @use(%v) : (f32) -> ()
-    }
-    return
+// COM: Ensure reductions loops guards are correct.  
+func.func @affine_for_hoist2(%arg0: memref<?xi32>, %arg1: i32) -> (i32) {
+  // CHECK:       #set = affine_set<() : (9 >= 0)>
+  // CHECK:        func.func @affine_for_hoist2(%arg0: memref<?xi32>, %arg1: i32) -> i32 {
+  // CHECK-NEXT:    %alloca = memref.alloca() : memref<1xi32>
+  // CHECK-NEXT:    %c3_i32 = arith.constant 3 : i32
+  // CHECK-NEXT:    %0 = affine.if #set() -> i32 {
+  // CHECK-NEXT:      %1 = affine.load %alloca[0] : memref<1xi32>
+  // CHECK-NEXT:      %2 = affine.for %arg2 = 0 to 10 iter_args(%arg3 = %c3_i32) -> (i32) {
+  // CHECK-NEXT:        %3 = arith.addi %arg3, %1 : i32
+  // CHECK-NEXT:        affine.yield %3 : i32
+  // CHECK-NEXT:      }
+  // CHECK-NEXT:      affine.yield %2 : i32
+  // CHECK-NEXT:    } else {
+  // CHECK-NEXT:      affine.yield %c3_i32 : i32
+  // CHECK-NEXT:    }
+  // CHECK-NEXT:    return %0 : i32
+
+  %alloca = memref.alloca() : memref<1xi32>    
+  %sum0 = arith.constant 3 : i32    
+  %sum = affine.for %arg2 = 0 to 10 iter_args(%sum_iter = %sum0) -> (i32) {    
+    %tmp = affine.load %alloca[0] : memref<1xi32>
+    %sum_next = arith.addi %sum_iter, %tmp : i32
+    affine.yield %sum_next : i32
   }
+  return %sum : i32
+}
```

Comment:
Can we have tests for the two other loop kinds we're currently handling?

Reply:
Not in this PR because in this PR we can handle reductions only on AffineForOp type of loops. 
In this PR we only handle:
   - AffineForOp with/without reduction
   - scf::ParallelOp only without reduction
   - AffineParallelOp only without reduction
 We have tests for those scenarios. Future Prs will add the functionality that is missing along with more tests.